### PR TITLE
update brain to actually know the rules

### PR DESCRIPTION
**Pull request by deebloo**
update the move maker to choose an available move regardless of past moves

### CODE REVIEW

**Comment by deebloo** [comment]
note: This is SIGNIFICANTLY increasing the training time

**Comment by zerobasedindex** [comment]
I find it odd that this `SIGNIFICANTLY` increases the time, can you profile this to figure out what's taking the most time?

**Comment by deebloo** [comment]
I haven't looked into it much yet. But there are several more iterations. That have to happened for each move. It's possible those milliseconds just add up but seems unlikely that could be everything.

**Comment by zerobasedindex** [comment]
How memory efficient is node/typescript with maps? A big changes is reading in a file and fully allocating memory for it, and then going to continually adding data to an object and letting it grow naturally, could lead to some memory trashing

Original
```
real    0m40.706s
user    0m40.578s
sys     0m0.172s
```

Your Branch:
```
real    2m10.433s
user    1m12.047s
sys     1m3.234s
```

I used `time` for the calculations, I stopped the second after 50000 rounds on a 100000 train. The big difference in timing for system points to us doing something drastically different, and the larger gap between user and real backs that up. The app isn't running as much, it's running other stuff to handle what it needs more.

I could see the need to allocate a new object, copy data around, and then garbage collect causing some trashing here compared to before when we just update fields in an already fleshed about map, no really memory trashing required. Thoughts? How does node handle updating objects with new attributes?

This line would be the one making the object bigger right? `this.brain.gameStates[move.key] = { moves };`

**Comment by zerobasedindex** [comment]
Try updating the code to save the brain state as an array and doing a filter instead of an attribute lookup. You would need to pre-allocate the array with a larger number, say `19683` so that it doesn't have to remake the array over and over.

**Comment by deebloo** [comment]
@zerobasedindex I can give it a shot. I don't think an array will help since the longer it runs the lookup will take longer. The other thing I can try to do is use a proper Map and save the value off the the json file. The reason I don't think an array will help is that in JS an array isn't really an array. and array is an object.

screen shot 2018-01-09 at 7 49 59 am

**Comment by deebloo** [comment]
@zerobasedindex oh shit I stand corrected

File: src/brain-map.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  createHash(key: string) {
+    let hash = 5381;
+    let i = key.length;
+
+    while (i) {
+      hash = (hash * 33) ^ key.charCodeAt(--i);
+    }
+
+    return (hash >>> 0) % 20000;
```
Comment (zerobasedindex) [inline src/brain-map.ts:22]: So I'm not familiar with the `>>>` operator, what does that do?

File: src/brain-map.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  state: GameState[] = new Array(20000);
+
+  constructor(initState: GameState[] = []) {
+    initState.forEach((move, i) => {
+      this.state[i] = move;
+    });
+  }
+
+  set(key: string, state: GameState) {
+    this.state[this.createHash(key)] = state;
```
Comment (zerobasedindex) [inline src/brain-map.ts]: What happens when two different keys result in the same hash value? The ole hash collision problem.

Comment (deebloo) [reply src/brain-map.ts]: I was thinking about it and with our limited dataset is a hash collision a real concern?

File: src/brain-map.ts (hunk 1 above)
Comment (deebloo) [reply src/brain-map.ts:22]: I honestly don't know. I took it from the another hash function implementation

Comment (zerobasedindex) [reply src/brain-map.ts:22]: Haha, okay, I know that `>>` is a bit shift. So it looks like `>>>` is a `0` fill bit shift to the right. So you are shifting the binary representation of hash to the right 0 bits. So nothing.

Example: 5 >>> 2 would be:
5: `00000101` shifting to the right two `00000000`

File: src/brain-map.ts (hunk 2 above)
Comment (zerobasedindex) [reply src/brain-map.ts]: It's not worth it to test the speed, but wouldn't trust it going forward. We had a collision hit between two short words in the Java string hash implementation at Groupon. So it does happen. But it wouldn't have an impact on testing the speed of the solution.

File: src/brain-map.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { GameState } from './game-state';
+
+export class BrainMap {
+  state: GameState[] = new Array(20000);
+
+  constructor(initState: GameState[] = []) {
+    initState.forEach((move, i) => {
```
Comment (zerobasedindex) [inline src/brain-map.ts]: What is this doing exactly and how long does it take?

File: src/brain-map.ts (hunk 1 above)
Comment (zerobasedindex) [reply src/brain-map.ts:22]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators

File: src/brain-map.ts (hunk 3 above)
Comment (deebloo) [reply src/brain-map.ts]: this won't take long and it only happens once per training run

this is setting the initial state from the stored brain and setting it to the state array

File: src/brain-map.ts (hunk 2 above)
Comment (deebloo) [reply src/brain-map.ts]: gotcha. ill make sure it is safe

**Pull request merged**